### PR TITLE
Document streaming body

### PR DESCRIPTION
**Pull request by rayluo**
Automatically document all operations with a streaming body. Previously they were incorrectly documented as bytes.

This fixes boto/boto3#330

### CODE REVIEW

**Comment by jamesls** [comment]
High level:

1. Can you show an example of the rendered docs?
2. I see there's functional tests, but I feel we should still have unit tests for this.

File: docs/source/reference/response.rst [hunk 1]
```diff
@@ -0,0 +1,11 @@
+.. _ref-response:
+
+==================
+Response Reference
+==================
+
+botocore.response
+-----------------
+
+.. automodule:: botocore.response
```
Comment (jamesls) [inline docs/source/reference/response.rst]: Won't this document any additional functions we add to this module?  I'd prefer we just call out StreamingBody for now.

File: botocore/response.py [hunk 2]
```diff
@@ -67,6 +67,11 @@ def set_socket_timeout(self, timeout):
             raise
 
     def read(self, amt=None):
+        """
+        Read at most amt bytes from the stream.
```
Comment (jamesls) [inline botocore/response.py]: This should be on the first line of the docstring, line 70 (pep257).

Comment (rayluo) [inline botocore/response.py]: Thanks! Will do.

File: tests/unit/docs/test_method.py [hunk 3]
```diff
@@ -301,3 +301,16 @@ def test_exclude_output(self):
...
+        self.json_model['shapes']['Blob'] = {'type': 'blob'}
+        self.json_model['shapes']['SampleOperationInputOutput']['payload'] = \
+            'Body'
+        document_model_driven_method(
+            self.doc_structure, 'foo', self.operation_model,
+            event_emitter=self.event_emitter,
+            method_description='This describes the foo method.',
+            example_prefix='response = client.foo'
+        )
+        self.assert_contains_line('StreamingBody')
```
Comment (jamesls) [inline tests/unit/docs/test_method.py]: Can we make this assertion check that it's associated with the right param name as well.  Similar to what the tests above this one do.

Comment (kyleknap) [inline tests/unit/docs/test_method.py]: +1

File: botocore/docs/shape.py [hunk 4]
```diff
@@ -19,10 +19,12 @@
 class ShapeDocumenter(object):
     EVENT_NAME = ''
 
-    def __init__(self, service_name, operation_name, event_emitter):
+    def __init__(self, service_name, operation_name, event_emitter,
+                 operation_model=None):
```
Comment (jamesls) [inline botocore/docs/shape.py]: Is there a reason this is an optional param?

**Comment by kyleknap** [comment]
Looks fine. It is sort of weird that we pass the ``operation_model`` to the constructor of ``ShapeDocumenter`` when we access properties of ``operation_model`` to pass to the constructor. But I think we can wait on that refactoring this. :ship:

File: botocore/docs/example.py [hunk 5]
```diff
@@ -12,9 +12,12 @@
 # language governing permissions and limitations under the License.
 from botocore.docs.shape import ShapeDocumenter
 from botocore.docs.utils import py_default
+from botocore.model import get_streaming_body
 
 
 class BaseExampleDocumenter(ShapeDocumenter):
+    streaming_shape = None
```
Comment (kyleknap) [inline botocore/docs/example.py]: I would prefer this is not an attribute of the class. Much rather have it live in the constructor as an attribute of a ``ShapeDocumenter`` instance. If you are worried that ``streaming_shape ``is too specific, you can create a context attribute that is a dictionary that is used to store a bunch of this sort of metadata.

**Comment by kyleknap** [comment]
I like this approach more. Just had one comment.

File: botocore/model.py [hunk 6]
```diff
@@ -321,6 +321,20 @@ def signature_version(self, value):
         self._signature_version = value
 
 
+def get_streaming_body(shape):
```
Comment (jamesls) [inline botocore/model.py]: Not a fan of having this be a module function.  This seems like a method that should live on a shape class.

File: botocore/docs/example.py [hunk 7]
```diff
@@ -48,6 +50,8 @@ def document_recursive_shape(self, section, shape, **kwargs):
     def document_shape_default(self, section, shape, history, include=None,
                                exclude=None, **kwargs):
         py_type = py_default(shape.type_name)
+        if self._context.get('streaming_shape') == shape:
```
Comment (mtdowling) [inline botocore/docs/example.py]: Is the caching of `streaming_shape` inside of a `self._context` value necessary? Do we need to figure out if it's streaming or not multiple times for a single shape?

Comment (rayluo) [inline botocore/docs/example.py]: You are right. It is only for documentation, and we won't call it too often. I have removed the cache in later commit.

**Comment by mtdowling** [comment]
This seems fine. :ship:

**Comment by JordonPhillips** [comment]
looks good :shipit:

File: botocore/docs/method.py [hunk 8]
```diff
@@ -209,6 +209,8 @@ def document_model_driven_method(section, method_name, operation_model,
         return_section.style.indent()
         return_section.style.new_line()
 
+        context = {'streaming_shape': operation_model.get_streaming_output()}
```
Comment (kyleknap) [inline botocore/docs/method.py]: I like this actually more than what I originally suggested. The only possible issue that I could see is if a streaming output shape is shared elsewhere in the output shape but is not actually a streaming output, but I am having trouble seeing this happening.

**Comment by kyleknap** [comment]
:ship: on the assumption that a streaming output shape will never be reused in the output shape as a non streaming output.

File: botocore/docs/method.py (hunk 8 above)
Comment (rayluo) [inline botocore/docs/method.py]: Now I remembered that I checked that during my preliminary research, and I did a manual check just now to double confirm that, for the following sample model definition where the 2 members share the same `blobType` shape, they will end up as 2 different shape instances despite they are with the same type. So there would be no problem when we later use `if self._context.get('streaming_shape') == shape`.

            'shapes': {
                'OperationNameResponse': {
                    'type': 'structure',
                    'members': {
                        "Blob": {'shape': 'blobType'},
                        "Body": {'shape': 'blobType'}
                    },
                    'payload': 'Body'
                },
                'blobType': {
                    'type': 'blob'
                }
            }

So, no need to worry. It is a good question. Thanks!

**Pull request merged**